sql/mf_iocache_encr.cc

```diff
@@ -0,0 +1,251 @@
+/*
+   Copyright (c) 2017, Percona Inc. All Rights Reserved.
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; version 2 of the License.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program; if not, write to the Free Software
+   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA
+*/
+
+/*************************************************************************
+  Limitation of encrypted IO_CACHEs
+  1. Designed to support temporary files only (open_cached_file, fd=-1)
+  2. Created with WRITE_CACHE, later can be reinit_io_cache'ed to
+     READ_CACHE and WRITE_CACHE in any order arbitrary number of times.
+  3. no seeks for writes, but reinit_io_cache(WRITE_CACHE, seek_offset)
+     is allowed (there's a special hack in reinit_io_cache() for that)
+*/
+
+#include "my_global.h"
+#include "my_sys.h"
+#include "my_aes.h"
+#include "my_rnd.h"
+#include "log.h"
+#include "mysql/psi/mysql_file.h"
+
+typedef struct {
+  ulonglong counter;
+  uint block_length, last_block_length;
+  uchar key[MY_AES_BLOCK_SIZE];
+  ulonglong inbuf_counter;
```

Comment:
Move after "counter" to save some alignment padding bytes?

Reply:
Moved, although in this particular case there is no padding gaps:
```c++
ulonglong counter; // 8 bytes
uint block_length, last_block_length; // 4 + 4 = 8 bytes
uchar key[MY_AES_BLOCK_SIZE]; // 16 bytes
ulonglong inbuf_counter; // 8 bytes
```